### ISSUE TITLE
When I apply cam to Semantic segmentation network（Two categories），it has an error：An exception occurred in CAM with block: <class 'IndexError'>. Message: index 1 is out of bounds for dimension 0 with size 1

### ISSUE THREAD

**Issue by ChenHongyu-bo** [jacobgil/pytorch-grad-cam#241]

**Comment by jacobgil** [comment]
Hi,
I need details to help with this.
Can you please share some code?
What are the target_layers, and the target functions you're using ?

**Comment by ChenHongyu-bo** [comment]
> Hi, I need details to help with this. Can you please share some code? What are the target_layers, and the target functions you're using ?

Hi, thank you for your reply
I have shown my code
The output of my network has only two types, one is the background and one is the target areas
It works when I set it as follows:
`   grayscale_cam = cam(input_tensor=input_tensor,targets=None,)[0, :]`
But when I do this, I get the error I had before and I'm confused
`    grayscale_cam = cam(input_tensor=input_tensor,targets=targets,)[0, :]`
My target layers is the last layer of the decoder layers of my network
The loss function is BCEWithLogitsLoss



`model_path='../Iou_0.781.pkl'
model = NewNet()
model.load_state_dict(torch.load(model_path))
model.eval()

image_path='0_0_25_256_256.png'
img=np.array(Image.open(image_path))
rgb_img = np.float32(img) / 255
input_tensor = preprocess_image(rgb_img,
                                mean=[0.485, 0.456, 0.406],
                                std=[0.229, 0.224, 0.225])

if torch.cuda.is_available():
    model = model.cuda()
    input_tensor = input_tensor.cuda()


res, o1, o2, o3, o4 = model(input_tensor)


output  = res .sigmoid().data.cpu().numpy().squeeze()


plopymaskuint8 = 255 * np.uint8(output  == 1)
plopymaskfloat = np.float32(output  == 1)


from pytorch_grad_cam import GradCAM

targets = [SemanticSegmentationTarget(1, plopymaskfloat)]
target_layers = [model.m]

with GradCAM(model=model,target_layers=target_layers,use_cuda=torch.cuda.is_available()) as cam:

    grayscale_cam = cam(input_tensor=input_tensor,targets=targets,)[0, :]

    cam_image = show_cam_on_image(rgb_img, grayscale_cam, use_rgb=True)
    Image.fromarray(cam_image)
`

**Comment by jacobgil** [comment]
Hi, sorry for the delay in the response.

It looks like the model has one output (representing two categories).
Therefore 
targets = [SemanticSegmentationTarget(1, maskfloat)]
Should be changed to
targets = [SemanticSegmentationTarget(0, maskfloat)]
Since the model has only one output, representing the score of the second category.

**Comment by ChenHongyu-bo** [comment]
Thank you for your advice, I did as you said, but it seems that there is an error, the image is all blue（My task is to detect shadows）
![MJNMT19J70TM)MML(`75 I2](https://user-images.githubusercontent.com/104547695/166939461-4e053c32-f3b2-4c80-aafd-fc87cd782be3.png)

**Comment by ChenHongyu-bo** [comment]
If I set the targets to None, it will work better

**Comment by ChenHongyu-bo** [comment]
Ohh Yes, I forget about this detail, thanks to your reminder，I have changed it to maskfloat = np.float32(output > 0 )  , because the final output is a binary plot, so I think after sigmoid, I just need to set it to > 0 and It works well in the end. Not sure if my understanding is correct, but I solved the problem ,thank you very much.
(maskuint8 = 255 * np.uint8(output > 0.5)  This I wrote according to the tutorial you gave, and did not use it later, I forgot to comment it out, sorry)

**Comment by jacobgil** [comment]
maskfloat = np.float32(output > 0 )
But why would this work with 0 and not 0.5.
If you use output > 0, the entire mask will be positive.

Don't you instead want to just a mask that get a higher scores for the second category,
which means 0.5 after sigmoid?

shouldn't it be
maskuint8 = 255 * np.uint8(output > 0.5) ?
